source/index.ts

```diff
@@ -0,0 +1,317 @@
+import EventEmitter from 'eventemitter3';
+import {Queue} from './queue';
+import PriorityQueue from './priority-queue';
+import {QueueAddOptions, DefaultAddOptions, Options} from './options';
+
+type ResolveFunction<T = void> = (value?: T | PromiseLike<T>) => void;
+
+type Task<TaskResultType> =
+		| (() => PromiseLike<TaskResultType>)
+		| (() => TaskResultType);
+
+// tslint:disable-next-line:no-empty
+const empty = () => {};
+
+/**
+ * Promise queue with concurrency control.
+ */
+export default class PQueue<Q extends Queue<EnqueueOptionsType>, EnqueueOptionsType extends QueueAddOptions = DefaultAddOptions> extends EventEmitter<'active'> {
+	private readonly carryoverConcurrencyCount: boolean;
+	private readonly isIntervalIgnored: boolean;
+	private intervalCount: number;
+	private readonly intervalCap: number;
+	private readonly interval: number;
+	private intervalEnd: number;
+	private intervalId?: NodeJS.Timeout;
+	private timeoutId?: NodeJS.Timeout;
+
+	private queue: Q;
+	private readonly queueClass: new () => Q;
+
+	private pendingCount: number;
+	private readonly concurrency: number;
+	private paused: boolean;
+
+	private resolveEmpty: ResolveFunction;
+	private resolveIdle: ResolveFunction;
+
+	constructor(opt?: Options<Q, EnqueueOptionsType>) {
+		super();
+
+		const options: Options<Q, EnqueueOptionsType> = {
+			carryoverConcurrencyCount: false,
+			intervalCap: Infinity,
+			interval: 0,
+			concurrency: Infinity,
+			autoStart: true,
+			// @ts-ignore This default is a little nasty!
```

Comment:
Why?

Reply:
the classic old typescript problem of passing in class constructors from generics without being able to pass in types because they don't really exist 🤷‍♂️ 
<img width="948" alt="Screenshot 2019-04-11 at 02 13 15" src="https://user-images.githubusercontent.com/8214559/55922098-fb27fa00-5bff-11e9-9757-faa191a9cfa4.png">
<img width="858" alt="Screenshot 2019-04-11 at 02 13 23" src="https://user-images.githubusercontent.com/8214559/55922099-fbc09080-5bff-11e9-968b-5fe556ae3148.png">

Reply:
@BendingBender Any ideas?

Reply:
fix for now: https://github.com/sindresorhus/p-queue/pull/60#discussion_r274335468